### PR TITLE
added perplexity as metric

### LINKED ISSUE #153: [FEATURE] Loss as metric
### 🚀 Feature

Add a separate metric that calculates the loss only.

### Motivation

Generate can take a long time, but is needed for current metrics. 
Having a quick option with loss can be useful.

Issue is that validation insights cannot be generated in that case though.

### PR DESCRIPTION
**Pull request by pascal-pfeiffer**
closes #153 

also fixes plot_batch by escaping html code

### CODE REVIEW

File: llm_studio/src/datasets/text_causal_language_modeling_ds.py [hunk 1]
```diff
@@ -201,35 +204,31 @@ def clean_output(
...
+            output = self.clean_output(output, self.prompts, cfg)
 
         output["target_text"] = self.answers
-        metric_func, _ = cfg.prediction.metric_class.get(cfg.prediction.metric)
-        predictions = output["predicted_text"]
-        labels = output["target_text"]
-        assert len(predictions) == len(labels)
+        if "predicted_text" not in output.keys():
+            # In perplexity metric, no predictions were made, use the target text
+            output["predicted_text"] = output["target_text"]
```
Comment (maxjeblick) [inline llm_studio/src/datasets/text_causal_language_modeling_ds.py:213]: I'd not set predicted text at all. I can confuse user who download predictions while being unaware that perplexity metric triggers no text generation.

**Review by maxjeblick — COMMENTED** [review]
Thanks a lot for this PR, adding metrics that do not require `.generate` is certainly helpful as it allows to speed up experiments significantly. 

Some comments:
- I'd inform the user (in the logs/validation insights) that Perplexity does not generate predicted answers.
- I think we need to rewrite some part of the code to remove the necessity to have `"predicted_text"` column in the output dictionary (e.g. when calling `plot_validation_predictions`). 
- I'd move the logic whether to generate text or not out of the model to the evaluation loop and pass it as an argument to the model (need to think how it's done best). E.g. replace `calculate_loss` (which could always be done, given the labels) by `generate_text` flag and set this to `False` whenever the given metric does not need it. The reason for this is that currently a model trained with perplexity will not generate text which probably breaks the chat ui, etc.
- Would it be possible to also add loss as metric, as well (for feature completeness)?

File: llm_studio/src/losses/classification_losses.py [hunk 2]
```diff
@@ -25,6 +26,24 @@ def forward(self, logits, labels):
         return self.loss_fn(shift_logits, shift_labels)
 
 
+class Perplexity(nn.Module):
```
[the remark this answers is not in the document]
Comment (maxjeblick) [reply llm_studio/src/losses/classification_losses.py:29]: Moving to metrics file sounds like a good solution.

**Comment by pascal-pfeiffer** [comment]
> Thanks a lot for this PR, adding metrics that do not require `.generate` is certainly helpful as it allows to speed up experiments significantly.
> 
> Some comments:
> 
> * I'd inform the user (in the logs/validation insights) that Perplexity does not generate predicted answers.

great suggestion, done

> * I think we need to rewrite some part of the code to remove the necessity to have `"predicted_text"` column in the output dictionary (e.g. when calling `plot_validation_predictions`).

done

> * I'd move the logic whether to generate text or not out of the model to the evaluation loop and pass it as an argument to the model (need to think how it's done best). E.g. replace `calculate_loss` (which could always be done, given the labels) by `generate_text` flag and set this to `False` whenever the given metric does not need it. The reason for this is that currently a model trained with perplexity will not generate text which probably breaks the chat ui, etc.

Great catch with the Chat!
I am already doing something in that direction in the RLHF PR, will do what is needed to make the chat work here. 

> * Would it be possible to also add loss as metric, as well (for feature completeness)?

Not doing this, as it is very unstable as a loss function (torch.exp)

**Comment by pascal-pfeiffer** [comment]
> Great catch with the Chat!
I am already doing something in that direction in the RLHF PR, will do what is needed to make the chat work here.

It's not an issue, as chat calls the generate function directly.

**Comment by pascal-pfeiffer** [comment]
> Great catch with the Chat!
I am already doing something in that direction in the RLHF PR, will do what is needed to make the chat work here.

It's not an issue, as chat calls the generate function directly.

**Review by maxjeblick — APPROVED** [review]
LGTM, thanks!
